src/Microsoft.OData.Core/SelectedPropertiesNode.cs

```diff
@@ -294,10 +294,19 @@ internal SelectedPropertiesNode GetSelectedPropertiesForNavigationProperty(IEdmS
             }
 
             // $select=Orders will include the entire subtree when there are no same expanded entity.
-            if (this.selectedProperties != null && this.selectedProperties.Contains(navigationPropertyName) &&
-                (this.children == null || !this.children.Any(n => n.Key.Equals(navigationPropertyName, StringComparison.Ordinal) && n.Value.isExpandedNavigationProperty)))

```

Comment:
This is the problem of using Linq lightly and excessively 🤦‍♂️

Any() is an inefficient O(n) operation when this can be done using O(1) since **children** is a dictionary. Not to mention the lambda + closure allocation, though that is a smaller issue than the O(n) cost.

        public static bool Any<TSource>(this IEnumerable<TSource> source)
        {
            if (source == null)
            {
                throw Error.ArgumentNull("source");
            }

            using (IEnumerator<TSource> enumerator = source.GetEnumerator())
            {
                if (enumerator.MoveNext())
                {
                    return true;
                }
            }

            return false;
        }

Reply:
Not sure why `Any()` was used in this case given that it's searching by key and `this.children` is dictionary.